nilearn/decomposition/dict_learning.py

```diff
@@ -0,0 +1,275 @@
+        """Compute the mask and the ICA maps across subjects
+
+        Parameters
+        ----------
+        imgs: list of Niimg-like objects
+            See http://nilearn.github.io/building_blocks/manipulating_mr_images.html#niimg.
+            Data on which PCA must be calculated. If this is a list,
+            the affine is considered the same for all.
+
+        confounds: CSV file path or 2D matrixf
+            This parameter is passed to nilearn.signal.clean. Please see the
+            related documentation for details
+        """
+        # Base logic for decomposition estimators
+        BaseDecomposition.fit(self, imgs)
+
+        if self.verbose:
+            print('[DictLearning] Loading data')
+        data = mask_and_reduce(self.masker_, imgs, confounds,
+                               reduction_ratio=self.reduction_ratio,
+                               n_components=self.n_components,
+                               random_state=self.random_state,
+                               memory_level=max(0, self.memory_level - 1),
+                               n_jobs=self.n_jobs,
+                               memory=self.memory)
+        if self.verbose:
+            print('[DictLearning] Learning initial components')
+        self._init_dict(data)
+
+        self._raw_fit(data)
+
+    def _raw_fit(self, data):
+        """Compute the mask and the maps across subjects, using raw_data. Can
+        only be called directly is dict_init and mask_img, or
+        components_init_ is provided
+
+        Parameters
+        ----------
+        data: ndarray,
+            Shape (n_samples, n_features)
+        """
+        n_samples, n_features = data.shape
+
+        if self.verbose:
+            print('[DictLearning] Computing initial loadings')
+        self._init_loadings(data)
+
+        dict_init = self._loadings_init
+
+        n_iter = ((n_features - 1) // 20 + 1) * self.n_epochs
+
+        if self.verbose:
+            print('[DictLearning] Learning dictionary')
+        self.components_, dictionary = self._cache(dict_learning_online,
+                                                   func_memory_level=2)(
+            data.T,
+            self.n_components,
+            alpha=self.alpha,
+            n_iter=n_iter,
+            batch_size=20,
+            method='cd',
+            dict_init=dict_init,
+            verbose=max(0, self.verbose - 1),
+            random_state=self.random_state,
+            return_code=True,
+            shuffle=True,
+            n_jobs=1)
+        self.components_ = self.components_.T
+        # Normalize components
+        S = np.sqrt(np.sum(self.components_ ** 2, axis=1))
+        S[S == 0] = 1
+        self.components_ /= S[:, np.newaxis]
+
+        # flip signs in each composant so that positive part is l1 larger
+        # than negative part
+        for component in self.components_:
+            if np.sum(component > 0) < np.sum(component < 0):
```

Comment:
@arthurmensch  has validated this choice empirically, and it works well.

@arthurmensch : could you add a comment in the code saying exactly this, so that we avoid people changing it in the future.